src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/RazorProjectInfoHelpers.cs

```diff
@@ -15,60 +19,45 @@
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Razor;
-using Microsoft.Extensions.Logging;
 
-namespace Microsoft.AspNetCore.Razor.ExternalAccess.RoslynWorkspace;
+namespace Microsoft.AspNetCore.Razor;
 
-internal static class RazorProjectInfoFactory
+internal static class RazorProjectInfoHelpers
 {
-    private static readonly StringComparison s_stringComparison;
-
-    static RazorProjectInfoFactory()
-    {
-        s_stringComparison = RuntimeInformation.IsOSPlatform(OSPlatform.Linux)
-            ? StringComparison.Ordinal
-            : StringComparison.OrdinalIgnoreCase;
-    }
-
-    public static async Task<RazorProjectInfo?> ConvertAsync(Project project, ILogger? logger, CancellationToken cancellationToken)
+    public static RazorProjectInfo? TryConvert(
+        Project project,
+        string projectPath,
+        string intermediateOutputPath,
+        ImmutableArray<TagHelperDescriptor> tagHelpers)
     {
-        var projectPath = Path.GetDirectoryName(project.FilePath);
-        if (projectPath is null)
-        {
-            logger?.LogInformation("projectPath is null, skip conversion for {projectId}", project.Id);
-            return null;
-        }
 
-        var intermediateOutputPath = Path.GetDirectoryName(project.CompilationOutputInfo.AssemblyPath);
-        if (intermediateOutputPath is null)
-        {
-            logger?.LogInformation("intermediatePath is null, skip conversion for {projectId}", project.Id);
-            return null;
-        }
-
-        // First, lets get the documents, because if there aren't any, we can skip out early
         var documents = GetDocuments(project, projectPath);
 
         // Not a razor project
         if (documents.Length == 0)
         {
-            if (project.DocumentIds.Count == 0)
-            {
-                logger?.LogInformation("No razor documents for {projectId}", project.Id);
-            }
-            else
-            {
-                logger?.LogTrace("No documents in {projectId}", project.Id);
-            }
-            
             return null;
         }
 
+        var options = project.AnalyzerOptions.AnalyzerConfigOptionsProvider;
+        var (razorConfiguration, rootNamespace) = ComputeRazorConfigurationOptions(options);
+
         var csharpLanguageVersion = (project.ParseOptions as CSharpParseOptions)?.LanguageVersion ?? LanguageVersion.Default;
+        var projectWorkspaceState = ProjectWorkspaceState.Create(tagHelpers, csharpLanguageVersion);
 
-        var options = project.AnalyzerOptions.AnalyzerConfigOptionsProvider;
-        var configuration = ComputeRazorConfigurationOptions(options, logger, out var defaultNamespace);
+        return new RazorProjectInfo(
+            projectKey: new ProjectKey(intermediateOutputPath),
```

Comment:
FYI that there's a `Project.ToProjectKey()` extension method somewhere that you could probably push down and avoid digging around for the intermediate output path.

Reply:
That looks to be in Workspace, should it be moved lower?

Reply:
IMO, it seems reasonable to move it `ProjectEngineHost`.